### ISSUE TITLE
high rate of churn for oc_mounts table

### ISSUE THREAD

**Issue by meonkeys** [nextcloud/server#35311]
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

---

## Steps to reproduce

```sql
SELECT MAX(id) FROM oc_mounts
```

## Issue description

I'm seeing what appears to be a massive amount of what I'll call churn against the `oc_mounts` table, and I think the root cause might be Nextcloud media/files shared in Talk. I've been using Talk for a while and have shared lots of media/files in chats.

This is either a bug report or performance enhancement idea.

Recently I experienced https://github.com/nextcloud/server/issues/33289 . Things improved when I upgraded to 24.0.7, but I'm still seeing lots of DML (mostly DELETEs and INSERTs) against `oc_mounts`.

I looked for slow queries by enabling the MariaDB slow query log with a threshold of 10 seconds. I have none.

I took row counts for all tables in my Nextcloud database. None are over 500k and most are much smaller.

I looked at `MAX(id)` for every table, and found that `oc_mounts` had only 7k rows, but the latest inserted rows had IDs greater than 6 million.

When I cleared all messages from a Talk chat with lots of attachments, I noticed the number of rows in `oc_mounts` decreased.

I don't have a stand-alone repro. The churn could be all due to https://github.com/nextcloud/server/issues/33289 , I don't know. I'm hoping this leads to (a) nothing, (b) a known design choice we can improve or (c) a bug we can fix.

If nextcloud/spreed isn't the correct project, maybe nextcloud/server is?

## Talk app

**Talk app version:** 14.0.6

**Custom Signaling server configured:** no

**Custom TURN server configured:** yes, (private / self-hosted coturn / seems to work OK)

**Custom STUN server configured:** yes, `stun.nextcloud.com:443`

## Server configuration

**Operating system**: Ubuntu 20.04 LTS 64-bit server

**Web server:** Apache

**Database:** MariaDB

**PHP version:** 8.0.25

**Nextcloud Version:** 24.0.7 (using Docker image)

**List of activated apps:**

<details>

```
  - accessibility: 1.10.0
  - activity: 2.16.0
  - admin_audit: 1.14.0
  - bruteforcesettings: 2.4.0
  - calendar: 3.5.2
  - circles: 24.0.1
  - cloud_federation_api: 1.7.0
  - comments: 1.14.0
  - contacts: 4.2.2
  - contactsinteraction: 1.5.0
  - cookbook: 0.9.18
  - dav: 1.22.0
  - deck: 1.7.2
  - event_update_notification: 2.0.0
  - federatedfilesharing: 1.14.0
  - files: 1.19.0
  - files_automatedtagging: 1.14.0
  - files_external: 1.16.1
  - files_fulltextsearch: 24.0.1
  - files_pdfviewer: 2.5.0
  - files_retention: 1.13.2
  - files_rightclick: 1.3.0
  - files_sharing: 1.16.2
  - files_trashbin: 1.14.0
  - files_versions: 1.17.0
  - files_videoplayer: 1.13.0
  - firstrunwizard: 2.13.0
  - forms: 2.5.1
  - fulltextsearch: 24.0.0
  - fulltextsearch_elasticsearch: 24.0.1
  - impersonate: 1.11.0
  - logreader: 2.9.0
  - lookup_server_connector: 1.12.0
  - maps: 0.2.1
  - metadata: 0.17.0
  - news: 19.0.0
  - nextcloud_announcements: 1.13.0
  - notes: 4.5.1
  - notifications: 2.12.1
  - oauth2: 1.12.0
  - password_policy: 1.14.0
  - passwords: 2022.11.21
  - phonetrack: 0.7.2
  - photos: 1.6.0
  - polls: 3.8.3
  - provisioning_api: 1.14.0
  - recommendations: 1.3.0
  - richdocuments: 6.3.1
  - serverinfo: 1.14.0
  - settings: 1.6.0
  - sharebymail: 1.14.0
  - spreed: 14.0.6
  - support: 1.7.0
  - systemtags: 1.14.0
  - tasks: 0.14.5
  - text: 3.5.1
  - twofactor_backupcodes: 1.13.0
  - twofactor_nextcloud_notification: 3.4.0
  - twofactor_totp: 6.4.1
  - twofactor_webauthn: 0.3.3
  - updatenotification: 1.14.0
  - user_status: 1.4.0
  - video_converter: 1.0.5
  - viewer: 1.8.0
  - workflowengine: 2.6.0
```
</details>

**Nextcloud configuration:**

<details>

```json
{
    "system": {
        "htaccess.RewriteBase": "\/",
        "htaccess.IgnoreFrontController": true,
        "memcache.local": "\\OC\\Memcache\\APCu",
        "apps_paths": [
            {
                "path": "\/var\/www\/html\/apps",
                "url": "\/apps",
                "writable": false
            },
            {
                "path": "\/var\/www\/html\/custom_apps",
                "url": "\/custom_apps",
                "writable": true
            }
        ],
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "nextcloud.example.com"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "24.0.7.1",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "3306",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "mail_smtpmode": "smtp",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_sendmailmode": "smtp",
        "mail_smtpport": "25",
        "trusted_proxies": "***REMOVED SENSITIVE VALUE***",
        "forwarded_for_headers": [
            "HTTP_X_FORWARDED_FOR",
            "HTTP_X_REAL_IP"
        ],
        "maintenance": false,
        "loglevel": 2,
        "ldapIgnoreNamingRules": false,
        "ldapProviderFactory": "OCA\\User_LDAP\\LDAPProviderFactory",
        "has_internet_connection": true,
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "password": "***REMOVED SENSITIVE VALUE***",
            "port": 6379
        },
        "enable_previews": true,
        "enabledPreviewProviders": [
            "OC\\Preview\\PNG",
            "OC\\Preview\\JPEG",
            "OC\\Preview\\GIF",
            "OC\\Preview\\BMP",
            "OC\\Preview\\XBitmap",
            "OC\\Preview\\MP3",
            "OC\\Preview\\TXT",
            "OC\\Preview\\MarkDown",
            "OC\\Preview\\Movie",
            "OC\\Preview\\PDF",
            "OC\\Preview\\SVG"
        ],
        "overwrite.cli.url": "https:\/\/nextcloud.example.com",
        "overwriteprotocol": "https",
        "app_install_overwrite": [
            "cookbook",
            "gpxedit"
        ],
        "default_phone_region": "US",
        "preview_max_x": "2048",
        "preview_max_y": "2048",
        "jpeg_quality": "60",
        "theme": "",
        "trashbin_retention_obligation": "auto, 30"
    }
}
```
</details>

### Server log (data/nextcloud.log)

Not applicable. At least, I'm not seeing any messages that look related to this issue.

**Comment by nickvergessen** [comment]
Moving to server. Talk does not do anything special, different or directly with mounts.
If there is an issue with that table it's a problem with the mount provider logic of server.
Seems like the other PR/issue you linked to from server is the hot trace. If that is still continuing it needs more fixing on the server

**Comment by solracsf** [comment]
I wonder if this can be somehow tied to https://github.com/nextcloud/server/issues/35044 🤔

**Comment by meonkeys** [comment]
@PVince81 Here are some stats for my `oc_mounts` table:

| statistic | query | result |
|------|-------|--------|
| max `id` value | `SELECT MAX(id) FROM oc_mounts` | 6914178 |
| row count | `SELECT COUNT(*) FROM oc_mounts` | 5438 |
| rows with null provider class | `SELECT COUNT(*) FROM oc_mounts WHERE mount_provider_class IS NULL` | 27 |

0.5% of the rows have NULL values for `mount_provider_class` (27 out of 5438).

The max `id` value seems to increase by about 100k-200k per day. I recall max `id` of about 6.6 million last week.

**Comment by PVince81** [comment]
wow 6914178 is quite high
how many users on your system ? how many entries in oc_filecache ?

it's not supposed to delete and insert the same entries repeatedly, if that's what is happening

@icewind1991 please help

**Comment by meonkeys** [comment]
> how many users on your system ?

4

> how many entries in oc_filecache ?

`SELECT COUNT(*) FROM oc_filecache` returns 507770.

**Comment by PVince81** [comment]
please try with 24.0.8 RC1 or 25.0.2 RC1 since there were some fixes for the mounts there

also an additional one here: https://github.com/nextcloud/server/pull/35518

**Comment by meonkeys** [comment]
Ah dang, this is in an instance I am using regularly, with real data. I (still) don't have a test sandbox nor a stand-alone / clean room reproduction of the bug. I'll see if I can figure out how to make a throw-away clone or stand-alone test, but more likely I'll just need to wait until the next release candidate is promoted to a release.

**Comment by meonkeys** [comment]
I'm still tracking this. Just upgraded to 25.0.2 and I see less DML against `oc_mounts` when I watch the MariaDB process list closely. Give me a few days and I should have some useful data to share.

**Comment by meonkeys** [comment]
I've collected about 7 days of data with 25.0.2.3 (and more with prior versions). Here's a snapshot. I'm not seeing any change in the average rate of churn across Nextcloud server versions.

[MAX(id) growth tracker.ods](https://github.com/nextcloud/server/files/10247563/MAX.id.growth.tracker.ods)

**Comment by meonkeys** [comment]
Would you mind adding the `25-feedback` label?

**Comment by meonkeys** [comment]
Is there any way to, like, reset all my mounts or something? Can I just `DELETE FROM oc_mounts`?

Sorry to be a pest, I'm just hitting this issue again and my server is unusable. Every web or desktop client request seems to trigger a DELETE and lots of INSERTs in `oc_mounts`.

**Comment by meonkeys** [comment]
I read a bunch of code. I ended up in the same place as this comment: https://github.com/nextcloud/server/issues/33289#issuecomment-1209925745 . I should do a clean room repro and crank up debugging while I attempt a fix or workaround, but I just don't have the time. I made sure my backups are good, then executed `DELETE FROM oc_mounts WHERE mount_provider_class IS NULL`. 27 rows deleted. Wish me luck. 😃

I'll keep an eye on the max id value in oc_mounts. Maybe the rate of growth for that value will slow down. I'll also of course keep an eye on performance.

**Comment by meonkeys** [comment]
I have thousands of rows in my `oc_mounts` table where `root_id` doesn't relate to any rows in `oc_filecache`. Is this expected?

Here's a few made-up rows that look like what I have about 4,700 of in my `oc_mounts` table:

| id | storage_id | root_id | user_id | mount_point | mount_id | mount_provider_class |
|----|----|----|----|----|----|----|
| 1994832 | 17 | 3502192 | fred | /fred/files/Photos/ | 4 | OCA\Files_External\Config\ConfigAdapter |
| 1994833 | 17 | 3502192 | bob | /bob/files/Photos/ | 4 | OCA\Files_External\Config\ConfigAdapter |
| 1994834 | 17 | 3502192 | joe | /joe/files/Photos/ | 4 | OCA\Files_External\Config\ConfigAdapter |

But there's no row in `oc_filecache` with a `fileid` of 3502192.

**Comment by meonkeys** [comment]
I cleaned up those orphan(?) rows with

```sql
DELETE FROM oc_mounts WHERE id IN (
     SELECT a.id
       FROM oc_mounts a
  LEFT JOIN oc_filecache b
         ON a.root_id = b.fileid
      WHERE b.fileid IS NULL
)
```

That deleted about 4,700 rows. My `oc_cache` table has about 900 rows now. I haven't noticed any negative effects yet. If anything, it feels a bit faster.

I noticed there is still some `DELETE` and `INSERT` action on `oc_mounts` when I open a fresh web browser to Nextcloud (root URL), so even though I don't understand the server code, I'm guessing mounts are related to a virtual filesystem in use during a user's session. Whatever is happening does seem to block the page loading. Subsequent page loads (and ajax requests) are very fast. I do not see this same slow-ish DML occurring against `oc_mounts` when I open up the Android app and browse files.

I'll keep an eye on `SELECT MAX(id) FROM oc_mounts` for at least a bit longer and share that again in case the data become useful.
